scalyr_agent/builtin_monitors/kubernetes_monitor.py

```diff
@@ -0,0 +1,1226 @@
+        for k,v in labels.iteritems():
+            flattened.append( k )
+            flattened.append( v )
+        md5.update( ''.join( flattened ) )
+
+        self.digest = md5.digest()
+
+
+class ContainerChecker( StoppableThread ):
+    """
+        Monitors containers to check when they start and stop running.
+    """
+
+    def __init__( self, config, logger, socket_file, docker_api_version, host_hostname, data_path, log_path ):
+
+        self._config = config
+        self._logger = logger
+
+        self.__delay = self._config.get( 'container_check_interval' )
+        self.__log_prefix = self._config.get( 'docker_log_prefix' )
+        name = self._config.get( 'container_name' )
+
+        self.__socket_file = socket_file
+        self.__docker_api_version = docker_api_version
+        self.__client = DockerClient( base_url=('unix:/%s'%self.__socket_file), version=self.__docker_api_version )
+
+        self.container_id = self.__get_scalyr_container_id( self.__client, name )
+
+        self.__checkpoint_file = os.path.join( data_path, "docker-checkpoints.json" )
+        self.__log_path = log_path
+
+        self.__host_hostname = host_hostname
+
+        self.__readback_buffer_size = self._config.get( 'readback_buffer_size' )
+
+        self.__glob_list = config.get( 'container_globs' )
+
+        self.containers = {}
+
+        self.__k8s = Kubernetes()
+
+        self.__log_watcher = None
+        self.__module = None
+        self.__start_time = time.time()
+        self.__thread = StoppableThread( target=self.check_containers, name="Container Checker" )
+
+    def start( self ):
+
+        self.containers = _get_containers(self.__client, ignore_container=self.container_id, glob_list=self.__glob_list, include_log_path=True)
+
+        # if querying the docker api fails, set the container list to empty
+        if self.containers == None:
+            self.containers = {}
+
+        k8s_data = self._get_k8s_data()
+        self.docker_logs = self.__get_docker_logs( self.containers, k8s_data )
+        self.raw_logs = []
+
+        #create and start the DockerLoggers
+        self.__start_docker_logs( self.docker_logs )
+        self._logger.log(scalyr_logging.DEBUG_LEVEL_1, "Initialization complete.  Starting docker monitor for Scalyr" )
+        self.__thread.start()
+
+    def stop( self, wait_on_join=True, join_timeout=5 ):
+        self.__thread.stop( wait_on_join=wait_on_join, join_timeout=join_timeout )
+
+        #stop the DockerLoggers
+
+        for logger in self.raw_logs:
+            path = logger['log_config']['path']
+            if self.__log_watcher:
+                self.__log_watcher.remove_log_path( self.__module, path )
+            self._logger.log(scalyr_logging.DEBUG_LEVEL_1, "Stopping %s" % (path) )
+
+        self.raw_logs = []
+
+    def _process_namespaces( self, namespaces ):
+        """ Iterates over all namespaces on this node and then queries
+        every pod in every namespace.
```

Comment:
Say something about the fact it's going to process all of the pods and what processing means.